File: src/resource/Arduino/MRLComm/MrlServo.cpp
<<<<<<< HEAD
#include <Servo.h>
#include "Msg.h"
#include "Device.h"
#include "MrlServo.h"

MrlServo::MrlServo(int deviceId) : Device(deviceId, DEVICE_TYPE_SERVO) {
  isMoving = false;
  isSweeping = false;
  // create the servo
  servo = new Servo();
  lastUpdate = 0;
  currentPos = 0.0;
  targetPos = 0;
  velocity = -1;
  acceleration = 1;
  moveStart = 0;
}

MrlServo::~MrlServo() {
  if (servo){
    servo->detach();
    delete servo;
  }
}

// this method "may" be called with a pin or pin & pos depending on
// config size
bool MrlServo::attach(byte pin, byte initPos, int initVelocity){
  // msg->publishDebug("MrlServo.deviceAttach !!!");
  servo->write(initPos);
  currentPos = initPos;
  targetPos = initPos;
  velocity = initVelocity;
  servo->attach(pin);
  return true;
}

// This method is equivalent to Arduino's Servo.attach(pin) - (no pos)
void MrlServo::attachPin(int pin){
  this->pin = pin;
  servo->attach(pin);
  servo->write((int)currentPos); //return to it's last know state (may be 0 if currentPos is not set)
  // TODO-KW: we should always have a moveTo for safety, o/w we have no idea what angle we're going to start up at.. maybe
}

void MrlServo::detachPin(){
  servo->detach();
}

// FIXME - what happened to events ?
void MrlServo::update() {
  //it may have an imprecision of +- 1 due to the conversion of currentPos to int
  if (isMoving) {
    if ((int)currentPos != targetPos) {
      long deltaTime = millis() - lastUpdate;
      float _velocity = velocity;
      if (acceleration != -1) {
        _velocity *= acceleration * pow(((float)(millis()- moveStart)) / 1000,2) / 10;
        //msg->publishDebug(String(_velocity));
        if (_velocity > velocity) {
          _velocity = velocity;
        }
      }
      float step = _velocity * deltaTime;
      step /= 1000; //for deg/ms;
      if (isSweeping) {
        step = sweepStep;
      }
      if (velocity < 0) { // when velocity < 0, it mean full speed ahead
        step = targetPos - currentPos;
      }
      else if ((int)currentPos > targetPos) {
        step *=-1;
      }
      currentPos += step;
      if ((step > 0.0 && (int)currentPos > targetPos) || (step < 0.0 && (int)currentPos < targetPos)) {
        currentPos = targetPos;
      }
      lastUpdate = millis();
      servo->write((int)currentPos);
    }
    else {
      if (isSweeping) {
        if (targetPos == min) {
          targetPos = max;
        }
        else {
          targetPos = min;
        }
        sweepStep *= -1;
      }
      else {
        isMoving = false;
      }
    }
  }
}

void MrlServo::servoWrite(int position) {
  if (servo == NULL) 
    return;
  targetPos = position;
  isMoving = true;
  lastUpdate = millis();
  moveStart = lastUpdate;
}

void MrlServo::servoWriteMicroseconds(int position) {
  if (servo) {
    servo->writeMicroseconds(position);
  }
}

void MrlServo::startSweep(int min, int max, int step) {
  this->min = min;
  this->max = max;
  sweepStep = step;
  targetPos = max;
  isMoving = true;
  isSweeping = true;
}

void MrlServo::stopSweep() {
  isMoving = false;
  isSweeping = false;
}

void MrlServo::setMaxVelocity(unsigned int velocity){
  maxVelocity = velocity;
}

void MrlServo::setVelocity(int velocity) {
  this->velocity = velocity;
}

void MrlServo::setAcceleration(int acceleration) {
  this->acceleration = acceleration;
}
=======
#include <Servo.h>
#include "Msg.h"
#include "Device.h"
#include "MrlServo.h"

MrlServo::MrlServo(int deviceId) : Device(deviceId, DEVICE_TYPE_SERVO) {
  isMoving = false;
  isSweeping = false;
  // create the servo
  servo = new Servo();
  lastUpdate = 0;
  currentPos = 0.0;
  targetPos = 0;
  velocity = -1;
  acceleration = -1;
  moveStart = 0;
}

MrlServo::~MrlServo() {
  if (servo){
    servo->detach();
    delete servo;
  }
}

// this method "may" be called with a pin or pin & pos depending on
// config size
bool MrlServo::attach(byte pin, byte initPos, int initVelocity){
  // msg->publishDebug("MrlServo.deviceAttach !!!");
  servo->write(initPos);
  currentPos = initPos;
  targetPos = initPos;
  velocity = initVelocity;
  servo->attach(pin);
  return true;
}

// This method is equivalent to Arduino's Servo.attach(pin) - (no pos)
void MrlServo::enablePwm(int pin){
  this->pin = pin;
  servo->attach(pin);
  servo->write((int)currentPos); //return to it's last know state (may be 0 if currentPos is not set)
  // TODO-KW: we should always have a moveTo for safety, o/w we have no idea what angle we're going to start up at.. maybe
}

void MrlServo::disablePwm(){
  servo->detach();
}

// FIXME - what happened to events ?
void MrlServo::update() {
  //it may have an imprecision of +- 1 due to the conversion of currentPos to int
  if (isMoving) {
    if ((int)currentPos != targetPos) {
      long deltaTime = millis() - lastUpdate;
      lastUpdate = millis();
      float _velocity = velocity;
      if (acceleration != -1) {
        _velocity *= acceleration * pow(((float)(millis()- moveStart)) / 1000,2) / 10;
        //msg->publishDebug(String(_velocity));
        if (_velocity > velocity) {
          _velocity = velocity;
        }
      }
      float step = _velocity * deltaTime;
      step /= 1000; //for deg/ms;
      if (isSweeping) {
        step = sweepStep;
      }
      if (velocity < 0) { // when velocity < 0, it mean full speed ahead
        step = targetPos - currentPos;
      }
      else if ((int)currentPos > targetPos) {
        step *=-1;
      }
      currentPos += step;
      if ((step > 0.0 && (int)currentPos > targetPos) || (step < 0.0 && (int)currentPos < targetPos)) {
        currentPos = targetPos;
      }
      servo->write((int)currentPos);
    }
    else {
      if (isSweeping) {
        if (targetPos == min) {
          targetPos = max;
        }
        else {
          targetPos = min;
        }
        sweepStep *= -1;
      }
      else {
        isMoving = false;
      }
    }
  }
}

void MrlServo::servoWrite(int position) {
  if (servo == NULL) 
    return;
  targetPos = position;
  isMoving = true;
  lastUpdate = millis();
  moveStart = lastUpdate;
}

void MrlServo::servoWriteMicroseconds(int position) {
  if (servo) {
    servo->writeMicroseconds(position);
  }
}

void MrlServo::startSweep(int min, int max, int step) {
  this->min = min;
  this->max = max;
  sweepStep = step;
  targetPos = max;
  isMoving = true;
  isSweeping = true;
}

void MrlServo::stopSweep() {
  isMoving = false;
  isSweeping = false;
}

void MrlServo::setMaxVelocity(unsigned int velocity){
  maxVelocity = velocity;
}

void MrlServo::setVelocity(int velocity) {
  this->velocity = velocity;
}

void MrlServo::setAcceleration(int acceleration) {
  this->acceleration = acceleration;
}
>>>>>>> branch 'develop' of https://github.com/MyRobotLab/myrobotlab.git

File: src/org/myrobotlab/arduino/virtual/MrlServo.java
package org.myrobotlab.arduino.virtual;

import org.myrobotlab.arduino.Msg;
import org.myrobotlab.logging.LoggerFactory;
import org.myrobotlab.service.VirtualArduino;
import org.myrobotlab.service.interfaces.Simulator;
import org.myrobotlab.virtual.VirtualServo;
import org.slf4j.Logger;


/**
 * copy pasted from MrlServo.h
 *
 */
public class MrlServo extends Device {
  
  public final static Logger log = LoggerFactory.getLogger(MrlServo.class);

	public static final int SERVO_EVENT_STOPPED = 1;
	public static final int SERVO_EVENT_POSITION_UPDATE = 2;
	
  VirtualServo servo; // servo pointer - in case our device is a servo
  public int pin;
  public boolean isMoving;
  boolean isSweeping;
  public int targetPosUs;
  public float currentPosUs;
  public int minUs;
  public int maxUs;
  public long lastUpdate;
  public int velocity; // in deg/sec  |  velocity < 0 == no speed control
  public int sweepStep;
  public int acceleration;
  public long moveStart;

  public MrlServo(int deviceId, VirtualArduino virtual) {
    super(deviceId, Msg.DEVICE_TYPE_SERVO, virtual);
  isMoving = false;
  isSweeping = false;
  // create the servo
  servo = new HardwareServo();
  lastUpdate = 0;
  currentPosUs = 0.0f;
  targetPosUs = 0;
  velocity = -1;
  acceleration = -1;
  moveStart = 0;
  }

/*
MrlServo::~MrlServo() {
  if (servo){
    servo->detach();
    delete servo;
  }
}
*/
  
// this method "may" be called with a pin or pin & pos depending on
// config size
boolean attach(int pin, int initPosUs, int initVelocity, String name){
      // msg->publishDebug("MrlServo.deviceAttach !");
  Simulator sim = virtual.getSimulator();
  if (sim != null){
    servo = sim.createVirtualServo(name);
  } else {
    servo = new HardwareServo();
  }
  servo.writeMicroseconds(initPosUs);
  currentPosUs = initPosUs;
  targetPosUs = initPosUs;
  velocity = initVelocity;
  servo.attach(pin);
  //publishServoEvent(SERVO_EVENT_STOPPED);
  return true;
  }

// This method is equivalent to Arduino's Servo.attach(pin) - (no pos)
public void attachPin(int pin) {
  log.info("servo id {}->attachPin({})", id, pin);
  this.pin = pin;
  servo.attach(pin);
}

public void detachPin() {
  log.info("servo id {}->detachPin()", id);
  servo.detach();
}

// FIXME - what happened to events ?
  public void update() {
    //it may have an imprecision of +- 1 due to the conversion of currentPosUs to int
    if (isMoving) {
      if ((int)currentPosUs != targetPosUs) {
        long deltaTime = millis() - lastUpdate;
        lastUpdate = millis();
        float _velocity = velocity;
        if (acceleration != -1) {
          _velocity *= acceleration * Math.pow(((float)(millis()- moveStart)) / 1000,2) / 10;
          //msg->publishDebug(String(_velocity));
          if (_velocity > velocity) {
            _velocity = velocity;
          }
        }
        if(targetPosUs > 500) {
          _velocity = map(_velocity, 0, 180, 544, 2400) - 544;
        }
        float step = _velocity * deltaTime;
        step /= 1000; //for deg/ms;
        if (isSweeping) {
          step = sweepStep;
        }
        if (velocity < 0) { // when velocity < 0, it mean full speed ahead
          step = targetPosUs - (int)currentPosUs;
        }
        else if (currentPosUs > targetPosUs) {
          step *=-1;
        }
        int previousCurrentPosUs = (int)currentPosUs;
        currentPosUs += step;
        if ((step > 0.0 && (int)currentPosUs > targetPosUs) || (step < 0.0 && (int)currentPosUs < targetPosUs)) {
          currentPosUs = targetPosUs;
        }
        if (!(previousCurrentPosUs == (int)currentPosUs)) {
          servo.writeMicroseconds((int)currentPosUs);
          if ((int)currentPosUs == targetPosUs) {
            publishServoEvent(SERVO_EVENT_STOPPED);
          }
          else {
            publishServoEvent(SERVO_EVENT_POSITION_UPDATE);
          }
        }
      }
      else {
        if (isSweeping) {
          if (targetPosUs == minUs) {
            targetPosUs = maxUs;
          }
          else {
            targetPosUs = minUs;
          }
          sweepStep *= -1;
        }
        else {
          isMoving = false;
          publishServoEvent(SERVO_EVENT_STOPPED);
        }
      }
    }
  }

 
 public void moveToMicroseconds(int posUs) {
    if (servo == null){ 
      return;
    }
    targetPosUs = posUs;
    isMoving = true;
    lastUpdate = millis();
    moveStart = lastUpdate;
    publishServoEvent(SERVO_EVENT_POSITION_UPDATE);
  }

void startSweep(int minUs, int maxUs, int step) {
  this.minUs = minUs;
  this.maxUs = maxUs;
  sweepStep = step;
  targetPosUs = maxUs;
  isMoving = true;
  isSweeping = true;
}

void stopSweep() {
  isMoving = false;
  isSweeping = false;
}

void setVelocity(int velocity) {
  this.velocity = velocity;
}
 
void setAcceleration(int acceleration) {
  this.acceleration = acceleration;
}

  private long millis() {
    return System.currentTimeMillis();
  }

  private void publishServoEvent(int type) {
    msg.publishServoEvent(id, type, (int)currentPosUs, targetPosUs);  
  }

  public void servoWriteMicroseconds(int position) {
  }

  private float map(float value, int inMin, int inMax, int outMin, int outMax) {
    return outMin + ((value - inMin) * (outMax - outMin)) / (inMax - outMax);
  }

};